src/model/Repository/Repository.php

```diff
@@ -43,7 +48,7 @@ public function __construct(
      */
     public function getTable()
     {
-        $database = $this->getDatabase(true);
```

Comment:
You changed the parameter to `$this->allowReplica`. I'm trying to understand _why_ and whether it's really necessary. 

It feels to like `$this->allowReplica` might be redundant. Our intention was to use `ReplicaManager::replicaConfig` as a single-point-of-truth flag to determine whether replication should be enabled or not.

- If `@replicaConfig` is not enabled in the `ReplicaManager`, then any call to `ReplicaManager::getDatabase()` (which is what gets called when you call `$this->getDatabase`) will fallback to the primary database and won't use replication at all.
- If it is enabled, than the replication is allowed and database selection is performed based on the previous queries and `ReplicaConfig` settings.

  By "enabled" I mean this:
  ```neon
  decorator:
	Crm\ApplicationModule\Repository:
		setup:
			- setReplicaConfig(@replicaConfig)
  ```

So the questions are:

- Why do you need `$this->allowReplica` here?
- Where do you want to call `ensure` from? What's the scenario? (since it's not called from anywhere in this PR)

Reply:
I'm calling `$this->getDatabase()` in the ensure method too, and wanted to use the same parameter value as in the `getTable()` method. It seemed easier like this, mainly because of future changes. If the value changes somehow, the `$this->allowReplica` property will change it in both places. But I could change the property to a constant, so the child repositories could not override it.

I want to use it in the payment card expire update command. Sometimes the gateway takes too long to download the card data, and the database connection gets timeouted, when the timeout threshold is low. If lifting the threshold is not an option, we could use this `ensure` method to reconnect if necessary.

Reply:
OK, I removed the `allowReplica` variable, but I'm not sure, if there has to be a true argument in tne `getDatabase` call in the `ensure` method. We use it mainly for reading, but it can be used for write as well. Yes, your example is correct, here is another one (both work):
```
$this->fooRepository->ensure(function(FooRepository $repository) use ($row) {
    $repository->update($row, ['foo' => 'bar']);
});
```

Reply:
Regarding the `true` argument in the `getDatabase` call, let me go through the scenarios:

##### `getDatabase(false)`

- Attempt to execute `$callback`.
- Failure in the DB connection, call `getDatabase(false)`
- CRM internally sets `write` flag for the requested table to `true` and all of the DB requests coming after this call would go to the primary database. This is because you requested the primary database and we must assume that DB requests after this need to go there too to maintain consitency.
- Second attempt to execute `$callback`, going to the just-reconnected primary database for the both reads and the writes.

The downside of this is that basically you'd force anything coming after this `ensure()` call to go to the primary server and ignore replica. Not sure if this is what you intended.

##### `getDatabase(true)`

- Attempt to execute `$callback`.
- Failure in the DB connection, call `getDatabase(true)`
- Reconnect to the database based on the previous requests/queries.
  - If there were only reads, we'd reconnect to the replica.
  - If there were writes, we'd reconnect to the primary.
- Second attempt to execute `$callback`, going to the reconnected instance.

I think this is more risky scenario. Let's consider this snippet:

```php
$this->fooRepository->ensure(function() use ($row) {
    $this->fooRepository->getTable()->fetch(); // read, going to the replica, failing
    $this->fooRepository->update($row, ['foo' => 'bar']); // write, going to the primary
});
```

In this case, we'd fail on the read from replica, reconnected the connection there and we could still again fail on the write to the primary if the connection is dropped there.

##### Conclusion

In the end, I think that the `getDatabse(false)` is the way to go here. It's the safer of the options. We'd just need the document the consequence that all further requests to the table will go through the primary database.

@mikoczy Are you OK with that?

Reply:
Yes, it looks good. ensure should be called mainly in commands, so connecting to the main db is not really a downside in my opinion..